indexing-service/src/main/java/org/apache/druid/indexing/common/task/CompactionTask.java

```diff
@@ -1014,6 +897,254 @@ static DimensionSchema createDimensionSchema(
+    {
+      if (!needQueryGranularity) {
+        throw new ISE("Not computing queryGranularity");
+      }
+
+      return queryGranularity;
+    }
+
+    public DimensionsSpec getDimensionsSpec()
+    {
+      if (!needDimensionsSpec) {
+        throw new ISE("Not computing dimensionsSpec");
+      }
+
+      final BiMap<Integer, String> orderedDims = uniqueDims.inverse();
+      final List<DimensionSchema> dimensionSchemas =
+          IntStream.range(0, orderedDims.size())
+                   .mapToObj(i -> {
+                     final String dimName = orderedDims.get(i);
+                     return Preconditions.checkNotNull(
+                         dimensionSchemaMap.get(dimName),
+                         "Cannot find dimension[%s] from dimensionSchemaMap",
+                         dimName
+                     );
+                   })
+                   .collect(Collectors.toList());
+
+      return new DimensionsSpec(dimensionSchemas);
+    }
+
+    public AggregatorFactory[] getMetricsSpec()
+    {
+      if (!needMetricsSpec) {
+        throw new ISE("Not computing metricsSpec");
+      }
+
+      if (aggregatorFactoryLists.isEmpty()) {
+        return new AggregatorFactory[0];
+      }
+
+      final AggregatorFactory[] mergedAggregators = AggregatorFactory.mergeAggregators(
+          aggregatorFactoryLists.stream()
+                                .map(xs -> xs.toArray(new AggregatorFactory[0]))
+                                .collect(Collectors.toList())
+      );
+
+      if (mergedAggregators == null) {
+        throw new ISE(
+            "Failed to merge existing aggregators when generating metricsSpec; "
+            + "try providing explicit metricsSpec"
+        );
+      }
+
+      return mergedAggregators;
+    }
+
+    /**
+     * Sort segments in order, such that we look at later segments prior to earlier ones. Useful when analyzing
+     * dimensions, as it allows us to take the latest value we see, and therefore prefer types from more recent
+     * segments, if there was a change.
+     */
+    private void sortSegmentsList()
+    {
+      segments.sort(
+          (o1, o2) -> Comparators.intervalsByStartThenEnd().compare(o2.lhs.getInterval(), o1.lhs.getInterval())
+      );
+    }
+
+    private void processRollup(final QueryableIndex index)
+    {
+      if (!needRollup) {
+        return;
+      }
+
+      // carry-overs (i.e. query granularity & rollup) are valid iff they are the same in every segment:
+      // Pick rollup value if all segments being compacted have the same, non-null, value otherwise set it to false
+      if (rollupIsValid) {
+        Boolean isRollup = index.getMetadata().isRollup();
+        if (isRollup == null) {
+          rollupIsValid = false;
+          rollup = false;
+        } else if (rollup == null) {
+          rollup = isRollup;
+        } else if (!isRollup.equals(rollup)) {
+          rollupIsValid = false;
+          rollup = false;
+        }
+      }
```

Comment:
Good idea. Updated to similar but equivalent:

```
final Boolean isIndexRollup = index.getMetadata().isRollup();
rollup = rollup && Boolean.valueOf(true).equals(isIndexRollup);
```

Reply:
Oops, that didn't quite work, since `rollup` starts as `null`. Changed it to this instead:

```
      final Boolean isIndexRollup = index.getMetadata().isRollup();

      if (rollup == null) {
        rollup = isIndexRollup;
      } else {
        rollup = rollup && Boolean.valueOf(true).equals(isIndexRollup);
      }
```